kv_connectors/llmd_fs_backend/src/csrc/storage/buffer.cpp

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2025 The llm-d Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// storage_buffer.cpp
+#include <torch/extension.h>
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <numa.h>
+#include <iostream>
+#include <fstream>
+#include <filesystem>
+#include <thread>
+#include <vector>
+#include <algorithm>
+#include <sys/sysinfo.h>
+
+#include "buffer.hpp"
+#include "debug_utils.hpp"
+
+// Thread-local buffer used by each IO thread
+thread_local StagingBufferInfo t_staging_buffer{};
+
+// Global buffers used by IO threads
+std::vector<StagingBufferInfo> g_staging_buffers;
+
+// Return thread-local staging buffer, allocating or reallocating if needed
+StagingBufferInfo get_thread_local_staging_buffer(size_t required_bytes) {
+    if (!t_staging_buffer.ptr || t_staging_buffer.size < required_bytes) {
+        if (t_staging_buffer.ptr) {
+            cudaFreeHost(t_staging_buffer.ptr);
+            t_staging_buffer.ptr = nullptr;
+            t_staging_buffer.size = 0;
+            std::cerr << "[WARN] Thread " << std::this_thread::get_id()
+                      << " existing staging buffer too small ("
+                      << t_staging_buffer.size << " bytes), reallocating "
+                      << required_bytes << " bytes\n";
+        }
+
+        size_t alloc_size = std::max(required_bytes, (size_t)16 * 1024 * 1024);
```

Comment:
nit: can we declare a `MIN_STAGING_BUFFER_SIZE` (or another appropriate name) constant?